### PR TITLE
pay: new LayerZero-based bridgers

### PR DESCRIPTION
**Pull request by alexhooketh**
WIP

### CODE REVIEW

File: packages/contract/script/DeployDaimoPayBridger.s.sol [hunk 1]
```diff
@@ -79,12 +81,17 @@ contract DeployDaimoPayBridger is Script {
...
+            msg.sender,
+            DEPLOY_SALT_LEGACY_MESH_BRIDGER
+        );
 
         console.log("cctpBridger address:", cctpBridger);
         console.log("cctpV2Bridger address:", cctpV2Bridger);
         console.log("acrossBridger address:", acrossBridger);
         console.log("axelarBridger address:", axelarBridger);
         console.log("hopBridger address:", hopBridger);
+        console.log("legacyMeshBridger address:", legacyMeshBridger);
```
Comment (dcposch) [inline packages/contract/script/DeployDaimoPayBridger.s.sol]: add USDT0 bridger

File: packages/contract/src/DaimoPayLayerZeroBridger.sol [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+import "./TokenUtils.sol";
+import "./interfaces/IDaimoPayBridger.sol";
+import {IOFT, SendParam, MessagingFee} from "@layerzerolabs/oft-evm/contracts/interfaces/IOFT.sol";
+
+/// @author Daimo, Inc
+/// @custom:security-contact security@daimo.com
+/// @notice Base contract for bridging via LayerZero OFT v2 with pluggable
+/// accounting and options.
+/// @dev Subclasses implement `_computeAccounting` to define how much to pull,
+/// what to send, guarantees, and any LZ options.
```
Comment (dcposch) [inline packages/contract/src/DaimoPayLayerZeroBridger.sol]: nice docs

File: packages/contract/src/DaimoPayLayerZeroBridger.sol [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        IOFT(route.app).send{value: fee.nativeFee}(sp, fee, refundAddress);
+
+        emit BridgeInitiated({
+            fromAddress: msg.sender,
+            fromToken: route.bridgeTokenIn,
+            fromAmount: accounting.inAmountLD,
+            toChainId: toChainId,
+            toAddress: toAddress,
+            toToken: route.bridgeTokenOut,
+            toAmount: desiredOutLD,
```
Comment (dcposch) [inline packages/contract/src/DaimoPayLayerZeroBridger.sol]: confirming toAmount is in units of toToken

File: packages/contract/src/DaimoPayLayerZeroBridger.sol [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            amountLD: accounting.sendAmountLD,
+            minAmountLD: accounting.minAmountLD,
+            extraOptions: accounting.extraOptions,
+            composeMsg: accounting.composeMsg,
+            oftCmd: accounting.oftCmd
+        });
+
+        // Quote LZ fee & ensure native coverage if paying in native.
+        MessagingFee memory fee = IOFT(route.app).quoteSend(sp, accounting.payInZRO); // standard OFT v2 path
+        require(accounting.payInZRO || address(this).balance >= fee.nativeFee, "insufficient native fee");
```
Comment (dcposch) [inline packages/contract/src/DaimoPayLayerZeroBridger.sol]: how would "pay in ZRO" work?

does DaimoPayLayerZeroBridger need to hold ZRO, and do we need to approve it IOFT?

File: packages/contract/src/DaimoPayLayerZeroBridger.sol [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    /// @param toChainId Destination EVM chain ID.
+    /// @param bridgeTokenOutOptions Candidate destination token/amount options.
+    /// @return route The configured route for `toChainId`.
+    /// @return outAmount Desired destination amount (local decimals).
+    function _resolveRouteAndOut(
+        uint256 toChainId,
+        TokenAmount[] calldata bridgeTokenOutOptions
+    ) internal view returns (LZBridgeRoute memory route, uint256 outAmount) {
+        route = bridgeRouteMapping[toChainId];
+        require(route.bridgeTokenOut != address(0), "route not found");
```
[the remark this answers is not in the document]
Comment (dcposch) [reply packages/contract/src/DaimoPayLayerZeroBridger.sol]: otoh, bridgeTokenOut should not be the native token -> we don't support that, only erc20

File: packages/contract/src/DaimoPayLayerZeroBridger.sol [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        /// @notice Amount sent via OFT (SendParam.amountLD).
+        uint256 sendAmountLD;
+        /// @notice Minimum amount guaranteed to arrive (SendParam.minAmountLD).
+        uint256 minAmountLD;
+        /// @notice LayerZero v2 options (SendParam.extraOptions).
+        bytes extraOptions;
+        /// @notice Optional compose message for OFT.
+        bytes composeMsg;
+        /// @notice Optional OFT command bytes.
+        bytes oftCmd;
```
[the remark this answers is not in the document]
Comment (alexhooketh) [reply packages/contract/src/DaimoPayLayerZeroBridger.sol]: yes 

an example: if we decide that stargate taxi is safe enough, we'll need to update the bridger to supply the taxi bit to oftCmd

a bunch of LZ bridgers use compose or oft

File: packages/contract/src/DaimoPayLayerZeroBridger.sol [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+    /// @notice Accounting policy returned by subclasses to parameterize accounting send.
+    /// @dev Decides how much to pull, what to send, receive guarantees, and LZ options.
+    struct Accounting {
+        /// @notice Amount pulled from the caller and approved to `app` (local decimals).
+        uint256 inAmountLD;
+        /// @notice Amount sent via OFT (SendParam.amountLD).
+        uint256 sendAmountLD;
+        /// @notice Minimum amount guaranteed to arrive (SendParam.minAmountLD).
+        uint256 minAmountLD;
```
[the remark this answers is not in the document]
Comment (dcposch) [reply packages/contract/src/DaimoPayLayerZeroBridger.sol]: so we have two different coins. at a glance
- `inAmountLD` = denominated in bridgeTokenIn
- `sendAmountLD` = not immediately obvious
- `minAmountLD` = not immediately obvious

why are inAmount and sendAmount ever different?

File: packages/contract/src/DaimoPayLayerZeroBridger.sol [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+            toChainId: toChainId,
+            toAddress: toAddress,
+            toToken: route.bridgeTokenOut,
+            toAmount: desiredOutLD,
+            refundAddress: refundAddress
+        });
+
+        if (address(this).balance > 0) {
+            // native coin refund
+            (bool success, ) = tx.origin.call{value: address(this).balance}(
```
Comment (andrewliu08) [inline packages/contract/src/DaimoPayLayerZeroBridger.sol]: `tx.origin` should be fine, but i would put the gasRefundAddress in the `extraData` param like the axelar bridger.

trying to think of edge cases where `tx.origin` would be bad...

File: packages/contract/test/DaimoPayLegacyMeshBridger.t.sol [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        opts[0] = TokenAmount({
+            token: IERC20(address(usdt)),
+            amount: 10_000_000 // 10 USDT
+        });
+
+        (, uint256 amountIn) = bridger.getBridgeTokenIn(DEST_CHAIN, opts);
+
+        // expected literal from 10M gross-up with 10 bps fee
+        uint256 expected = 10_010_011;
+        assertEq(amountIn, expected);
```
Comment (andrewliu08) [inline packages/contract/test/DaimoPayLegacyMeshBridger.t.sol]: how is this different from the above test?

File: packages/contract/src/DaimoPayLayerZeroBridger.sol [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    /// @param r Resolved LayerZero route for the destination.
+    /// @param desiredOutLD Desired receive amount on the destination (local decimals).
+    /// @param extraData Opaque data for subclass-specific policy.
+    /// @return a The computed accounting policy for the send.
+    function _computeAccounting(
+        uint256 toChainId,
+        address toAddress,
+        LZBridgeRoute memory r,
+        uint256 desiredOutLD,
+        bytes memory extraData
```
[the remark this answers is not in the document]
Comment (andrewliu08) [reply packages/contract/src/DaimoPayLayerZeroBridger.sol]: which other bridges would use extraData?

File: packages/contract/src/DaimoPayUSDT0Bridger.sol [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    /// @inheritdoc DaimoPayLayerZeroBridger
+    function _computeAccounting(
+        uint256 /*toChainId*/,
+        address /*toAddress*/,
+        LZBridgeRoute memory /*r*/,
+        uint256 desiredOutLD,
+        bytes memory /*extraData*/
+    ) internal pure override returns (Accounting memory a) {
+        // Plain OFT path: exact-out, no transfer bps.
+        a.inAmountLD = desiredOutLD; // pull from user
```
[the remark this answers is not in the document]
Comment (alexhooketh) [reply packages/contract/src/DaimoPayUSDT0Bridger.sol]: charging 3 bps for something that costs ~0 bps brings a whole lot of money

File: packages/contract/src/DaimoPayLayerZeroBridger.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrewliu08) [reply packages/contract/src/DaimoPayLayerZeroBridger.sol]: i don't think we'll bridge native token, but using eid feels sounder
